### PR TITLE
Promote CSINodeExpandSecret to GA

### PR DESCRIPTION
**Pull request by humblec**
This feature is in Beta since 1.27 and this has been promoted to GA in this release.

KEP: https://github.com/kubernetes/enhancements/pull/4239

Reference # https://github.com/kubernetes/enhancements/pull/4239

#### What type of PR is this?

/kind feature

-->

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
CSINodeExpandSecret feature has been promoted to GA in this release and enabled by default. The CSI drivers can make use of the `secretRef` values passed in NodeExpansion request optionally sent by the CSI Client from this release onwards. 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by swatisehgal — COMMENTED** [review]
There are a few place where we have checks to determine if this feature gate is enabled or not and we need to remove those as the feature gate is locked to default. Here are a few examples:

https://github.com/kubernetes/kubernetes/blob/d22e315c4ae45b293a5af38f3c2e387fcd8d8631/pkg/volume/csi/expander.go#L86

https://github.com/kubernetes/kubernetes/blob/64799e076097cf3c3dd12a82f9593a9bd57f042e/pkg/api/persistentvolume/util.go#L37

https://github.com/kubernetes/kubernetes/blob/64799e076097cf3c3dd12a82f9593a9bd57f042e/pkg/volume/csi/expander_test.go#L120

https://github.com/kubernetes/kubernetes/blob/64799e076097cf3c3dd12a82f9593a9bd57f042e/pkg/api/persistentvolume/util_test.go#L92

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -168,7 +168,7 @@ const (
 	// kep: https://kep.k8s.io/3171
 	// alpha: v1.25
 	// beta: v1.27
-	//
+	// GA: 1.29
```
Comment (xing-yang) [inline pkg/features/kube_features.go:171]: v1.29

**Comment by xing-yang** [comment]
Comments in the API needs to be updated from "beta" to "GA":
https://github.com/kubernetes/kubernetes/blob/v1.29.0-alpha.2/pkg/apis/core/types.go#L1808

**Comment by zhucan** [comment]
> kubernetes/pkg/api/persistentvolume/util_test.go

for util_test.go, can not remove the code, because of this test case:

**Comment by zhucan** [comment]
@swatisehgal 
image
and for this func, we can not remove it, because of it will cause the csi e2e test failed, because of it will drop the  secret.

**Comment by zhucan** [comment]
@swatisehgal  locked doesn't means not check, locked means can not modify, to my understand.

**Comment by humblec** [comment]
@swatisehgal is it mandatory to have this feature `lockedDefault` to `true` ? because I see there are many GA features without `lockingDefault to true` . Also looking at the test cases in place if we go with removal of the checks and lockedDefault true, this need good amount of refactoring of test cases. So, it seems that, its better to have those checks in place and for enabled default cases it's a NOOP anyway.

**Comment by humblec** [comment]
> @swatisehgal is it mandatory to have this feature `lockedDefault` to `true` ? because I see there are many GA features without `lockedDefault to true` . Also, looking at the test cases in place, if we go with removal of the feature gate checks and lockedDefault true, it need good amount of refactoring of existing test cases. In short, it seems that, its better to have those checks in place and for enabled default cases it's a NOOP anyway.

- [x] On last refresh of the patch, we have kept `lockedDefault` to `true`
- [x] The feature gate check has been removed except one place for the same reason of test case refactoring it require
- [x] All other comments are addressed and tests are passing.

**Comment by MaryamTavakkoli** [comment]
Hello!
Bug triage for the 1.29 release cycle is here!
The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)). as we want to ensure that each PR has a chance to be merged.

Are we still targeting the 1.29 release?

**Comment by xing-yang** [comment]
It still says "beta" in types.go:

https://github.com/kubernetes/kubernetes/blob/v1.29.0-alpha.2/pkg/apis/core/types.go#L1808
```
	// This is a beta field which is enabled default by CSINodeExpandSecret feature gate.
```

It still checks feature gate in the following places:
```
api/persistentvolume/util.go:	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeExpandSecret) && !hasNodeExpansionSecrets(oldPVSpec) {

volume/csi/expander_test.go:		enableCSINodeExpandSecret bool
volume/csi/expander_test.go:			enableCSINodeExpandSecret: true,
```

**Comment by zhucan** [comment]
```
volume/csi/expander_test.go:		enableCSINodeExpandSecret bool
volume/csi/expander_test.go:			enableCSINodeExpandSecret: true,
```
To my undestands, I think it means whether we need to create secrets, not means check featuregates.

**Comment by zhucan** [comment]
`api/persistentvolume/util.go:	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeExpandSecret) && !hasNodeExpansionSecrets(oldPVSpec) {`

for this, with e2e test, when creating pvc  with secret, if the oldpv.csi.spec is nil, but the newpv.csi.spec is not nil, when not checking the featuregates, the secret info will be dropped.

**Comment by xing-yang** [comment]
>```
>api/persistentvolume/util.go:	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeExpandSecret) && >!hasNodeExpansionSecrets(oldPVSpec) {
>```

>for this, with e2e test, when creating pvc with secret, if the oldpv.csi.spec is nil, but the newpv.csi.spec is not nil, if not checking the featuregates, the secret info will be dropped.

You can remove the following function from "api/persistentvolume/util.go" as you can no longer disable the feature gate :
```
// DropDisabledSpecFields removes disabled fields from the pv spec.
// This should be called from PrepareForCreate/PrepareForUpdate for all resources containing a pv spec.
func DropDisabledSpecFields(pvSpec *api.PersistentVolumeSpec, oldPVSpec *api.PersistentVolumeSpec) {
        if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeExpandSecret) && !hasNodeExpansionSecrets(oldPVSpec) {
                if pvSpec.CSI != nil {
                        pvSpec.CSI.NodeExpandSecretRef = nil
                }
        }
}
```

**Comment by zhucan** [comment]
> DropDisabledSpecFields

good idea

**Comment by zhucan** [comment]
> > ```
> > api/persistentvolume/util.go:	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeExpandSecret) && >!hasNodeExpansionSecrets(oldPVSpec) {
> > ```
> 
> > for this, with e2e test, when creating pvc with secret, if the oldpv.csi.spec is nil, but the newpv.csi.spec is not nil, if not checking the featuregates, the secret info will be dropped.
> 
> You can remove the following function from "api/persistentvolume/util.go" as you can no longer disable the feature gate :
> 
> ```
> // DropDisabledSpecFields removes disabled fields from the pv spec.
> // This should be called from PrepareForCreate/PrepareForUpdate for all resources containing a pv spec.
> func DropDisabledSpecFields(pvSpec *api.PersistentVolumeSpec, oldPVSpec *api.PersistentVolumeSpec) {
>         if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeExpandSecret) && !hasNodeExpansionSecrets(oldPVSpec) {
>                 if pvSpec.CSI != nil {
>                         pvSpec.CSI.NodeExpandSecretRef = nil
>                 }
>         }
> }
> ```

good idea

**Comment by xing-yang** [comment]
/assign @msau42 
for API approval

**Comment by humblec** [comment]
@wojtek-t can you please review/approve it from `registry/owners` side ?

**Comment by msau42** [comment]
Trying to get #121210 merged

**Comment by dims** [comment]
@humblec please rebase!

**Comment by humblec** [comment]
@xing-yang @msau42 rebased to master to resolve the conflict.. can you please bring back lgtm ?

**Comment by humblec** [comment]
> @humblec please rebase!

sure, done @dims

**Comment by humblec** [comment]
Thanks everyone  🎉  🎉

**Pull request merged**